dhall/ghc-src/Dhall/Import/HTTP.hs

```diff
@@ -238,20 +250,35 @@ corsCompliant (Remote parentURL) childURL responseHeaders = liftIO $ do
                 Control.Exception.throwIO (NotCORSCompliant {..})
 corsCompliant _ _ _ = return ()
 
-type HTTPHeader = Network.HTTP.Types.Header
+addHeaders :: OriginHeaders -> Maybe [HTTPHeader] -> HTTP.Request -> HTTP.Request
+addHeaders originHeaders urlHeaders request =
+    request { HTTP.requestHeaders = (filterHeaders urlHeaders) <> perOriginHeaders }
+      where
+        origin = decodeUtf8 (HTTP.host request) <> ":" <> Text.pack (show (HTTP.port request))
+ 
+        perOriginHeaders = HashMap.lookupDefault [] origin originHeaders
+
+        filterHeaders = foldMap (filter (not . overridden))
+
+        overridden :: HTTPHeader -> Bool
+        overridden (key, _value) = any (matchesKey key) perOriginHeaders
+
+        matchesKey :: CI ByteString -> HTTPHeader -> Bool
+        matchesKey key (candidate, _value) = key == candidate
 
 fetchFromHttpUrl :: URL -> Maybe [HTTPHeader] -> StateT Status IO Text.Text
 fetchFromHttpUrl childURL mheaders = do
+    Status { _loadOriginHeaders } <- State.get
+
+    originHeaders <- _loadOriginHeaders
```

Comment:
I made one mistake when I previously suggested using the ambient `StateT` for `_loadOriginHeaders`.  We do still want `_loadOriginHeaders` to use `StateT`, but we want the `_stack` to be reset to contain just one import, which is the path to origin headers.  This is because of this line from the standard:

```
(ε, headersPath) × Γ₀ ⊢ userHeadersExpr ⇒ userHeaders ⊢ Γ₁ ; Resolve userHeadersExpr with an empty import context
```

The key bit is the `(ε, headersPath)` part, which says that when resolving the `userHeadersExpr` you want the stack to contain only the headers path and nothing else.

However, we don't have to reset the stack here within the `fetchFromHttpUrl` function.  Instead, I think the best place to do that would be inside of the `_loadOriginHeaders` function.  I'll comment below with some more notes about how to do that.